### PR TITLE
JA: add site/ja/xla/custom_call.md

### PR DESCRIPTION
**Pull request by kiszk**
The original English file is at https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/g3doc/custom_call.md

### CODE REVIEW

File: site/ja/xla/custom_call.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+警告： カスタムコールのAPI/ABIは、現時点では固まっていません。
+きまぐれに変更するつもりはありませんが、変更する可能性はあります。
+将来可能性があるいくつかの変更については以下で説明します。
+
+## CPUでのカスタムコール
+
+XLAクライアントAPI経由で、カスタムコールを表すHLO命令を作ることができます。
+これは、執筆時点ではTensorFlow経由では公開されていません。
+
+例えば、以下のコードはCPU上で `A[i] = B[i % 128] + C[i]` をカスタム呼び出しを使用して計算します（もちろん、通常のHLOを使って計算できますし、すべきです！）。
```
[the remark this answers is not in the document]
Comment (kiszk) [reply site/ja/xla/custom_call.md]: Thanks, done

File: site/ja/xla/custom_call.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+  const float* in0 = reinterpret_cast<const float*>(in[0]);
+  const float* in1 = reinterpret_cast<const float*>(in[1]);
+  for (int i = 0; i < 2048; ++i) {
+    out_buf[i] = in0[i % 128] + in1[i];
+  }
+}
+XLA_REGISTER_CUSTOM_CALL_TARGET(do_custom_call, "Host");
+```
+
+関数 `do_custom_call` は、バッファの次元情報を知っている必要があります。
```
[the remark this answers is not in the document]
Comment (kiszk) [reply site/ja/xla/custom_call.md]: Good catch, done

File: site/ja/xla/custom_call.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+パラメータが最初に来て、そのあと出力の値が来ます。
+これは、CPUの呼び出し規則とは大きく異なり、２つのパラメータ、`ins` と `out` があります。
+違う実装をした主な理由は、タプル型の入出力を効率的に処理するためです。
+以下の章をごらんください。
+
+CPUの例のように、入出力バッファの大きさをカスタムコールに直書きしました。
+しかし、CPUの場合とは異なり、オペランドとしてバッファの大きさを渡してもうまく動きません。
+通常、CPU上でバッファの大きさが分かっている必要があります。例えば、カーネルを起動するとき、block/gridの次元情報が必要です。
+しかし、カスタムコールにオペランドとしてバッファサイズが渡されると、この値はGPUメモリ上にあります。
+処理の開始時に、この値を読むためのだけに重い同期的なデバイスからホストへのメモリコピーを実行する必要があります。
```
[the remark this answers is not in the document]
Comment (kiszk) [reply site/ja/xla/custom_call.md]: I use your suggestion, done.

File: site/ja/xla/custom_call.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+(*p0)[0] = subbuf0;
+(*p0)[1] = subtuple;
+(*p0)[2] = subbuf3;
+```
+
+CPUとGPUでメモリ内表現は同じですが、CPUとGPUのカスタムコール呼び出し規約では処理方法が異なります。
+
+### 一時バッファとしてのタプル出力
+
+カスタムコールへのタプル入力は便利ですが、必須ではありません。
```
[the remark this answers is not in the document]
Comment (kiszk) [reply site/ja/xla/custom_call.md]: Sure, done

File: site/ja/xla/custom_call.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+カスタムコールをつくるときに、任意のバイト文字列をセットできます。
+
+```c++
+std::string opaque = "...";
+xla::CustomCall(&b, "do_custom_call", /*operands=*/{param0, param1},
+                /*output_shape=*/ShapeUtil::CreateShape(F32, {2048}),
+                opaque);
+```
+
+`xla::Shape` はプロトコルバッファ表現を持つので、 `opaque` の内部にこれを保存してGPUカスタムコールの内部でデシリアライズできます。
```
[the remark this answers is not in the document]
Comment (kiszk) [reply site/ja/xla/custom_call.md]: Good catch. I used `表現` for the translation of `proto`. I think that `プロト` is not easy to understand.

File: site/ja/xla/custom_call.md [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+カスタムコールへのタプル入力は便利ですが、必須ではありません。
+カスタムコールへのタプル入力がサポートされていないなら、カスタムコールにタプルを渡す前にget-tuple-elementを使ってタプルを分解できます。
+
+一方、タプル*出力*は、他の方法ではできないことができます。
+
+タプル出力を持つ明確な理由は、それがカスタムコール（または、他のXLA命令）が複数の独立な配列を返す方法だからです。
+
+しかし、あまり明確ではないですが、タプル出力はカスタムコールに一時メモリを提供する方法でもあります。
+ええ、*出力*は一時バッファを表現できます。
+出力バッファは命令が書き込めるという性質を持っていて、書き込まれたあとに読み出すことができます。
```
[the remark this answers is not in the document]
Comment (kiszk) [reply site/ja/xla/custom_call.md]: Thanks. I use `オペレーション`.

**Comment by kiszk** [comment]
@nuka137 Thank you. I addressed all of your comments.

**Comment by nuka137** [comment]
LGTM. Thanks!

**Pull request merged**